File: ue5-sandbox 5.6/Source/Evo4XSandbox/MapGen/HexGrid.cpp
#include "HexGrid.h"
#include "SimCore/EvoRng.h"

FVector2D FHexGrid::HexToWorld(const FHexCoord& Coord)
{
	// Flat-top hex layout
	const float X = HexSize * (3.f / 2.f * Coord.Q);
	const float Y = HexSize * (FMath::Sqrt(3.f) * (Coord.R + Coord.Q / 2.f));
	return FVector2D(X, Y);
}

int32 FHexGrid::GetOrCreateTile(const FHexCoord& Coord)
{
	if (const int32* Existing = CoordToIndex.Find(Coord))
	{
		return *Existing;
	}

	const int32 Idx = Tiles.Num();
	FHexTile Tile;
	Tile.Index = Idx;
	Tile.Coord = Coord;
	Tile.WorldPos = HexToWorld(Coord);
	Tiles.Add(MoveTemp(Tile));
	CoordToIndex.Add(Coord, Idx);
	return Idx;
}

TArray<int32> FHexGrid::GetTileNeighbors(int32 TileIndex) const
{
	TArray<int32> Result;
	if (!Tiles.IsValidIndex(TileIndex)) return Result;

	const FHexCoord& Coord = Tiles[TileIndex].Coord;
	TArray<FHexCoord> Nbs = Coord.Neighbors();

	for (const FHexCoord& Nb : Nbs)
	{
		if (const int32* Idx = CoordToIndex.Find(Nb))
		{
			Result.Add(*Idx);
		}
	}
	return Result;
}

void FHexGrid::GenerateContinent(int32 TargetLandCount, const FString& Seed)
{
	Tiles.Empty();
	CoordToIndex.Empty();
	LandIndices.Empty();

	// Start from center
	const FHexCoord Center(0, 0);
	const int32 CenterIdx = GetOrCreateTile(Center);
	Tiles[CenterIdx].bIsLand = true;
	LandIndices.Add(CenterIdx);

	// Frontier for flood fill expansion
	TArray<int32> Frontier;
	Frontier.Add(CenterIdx);

	int32 Iteration = 0;

	while (LandIndices.Num() < TargetLandCount && Frontier.Num() > 0)
	{
		// Pick a frontier tile to expand from
		// Weight toward tiles closer to center for organic shape
		const FString PickKey = FString::Printf(TEXT("continent|pick|%d"), Iteration);
		const int32 FrontierIdx = FEvoRng::ChooseIndex(Seed, PickKey, Frontier.Num());
		const int32 TileIdx = Frontier[FrontierIdx];
		const FHexCoord& TileCoord = Tiles[TileIdx].Coord;

		// Try to expand to a random neighbor
		TArray<FHexCoord> Nbs = TileCoord.Neighbors();
		const FString NbKey = FString::Printf(TEXT("continent|nb|%d"), Iteration);
		const int32 NbPick = FEvoRng::ChooseIndex(Seed, NbKey, Nbs.Num());
		const FHexCoord& NbCoord = Nbs[NbPick];

		// Distance check: don't grow too far from center (creates organic blob)
		const int32 Dist = FHexCoord::Distance(Center, NbCoord);
		const int32 MaxRadius = FMath::CeilToInt32(FMath::Sqrt(static_cast<float>(TargetLandCount)) * 1.2f);

		// Probability of accepting decreases with distance (noise-like shape)
		const FString AcceptKey = FString::Printf(TEXT("continent|accept|%d"), Iteration);
		const float AcceptRoll = FEvoRng::Rand01(Seed, AcceptKey);
		const float AcceptChance = FMath::Clamp(1.f - static_cast<float>(Dist) / MaxRadius, 0.1f, 1.f);

		Iteration++;

		if (Dist > MaxRadius + 2)
		{
			// Too far, remove from frontier if all neighbors are land or too far
			continue;
		}

		if (AcceptRoll > AcceptChance)
		{
			continue; // Rejected by noise
		}

		// Create or find the neighbor tile
		const int32 NbIdx = GetOrCreateTile(NbCoord);

		if (!Tiles[NbIdx].bIsLand)
		{
			Tiles[NbIdx].bIsLand = true;
			LandIndices.Add(NbIdx);
			Frontier.Add(NbIdx);
		}

		// Remove frontier tile if all its neighbors are land
		bool bAllNeighborsLand = true;
		TArray<FHexCoord> CheckNbs = TileCoord.Neighbors();
		for (const FHexCoord& CheckNb : CheckNbs)
		{
			const int32* CheckIdx = CoordToIndex.Find(CheckNb);
			if (!CheckIdx || !Tiles[*CheckIdx].bIsLand)
			{
				bAllNeighborsLand = false;
				break;
			}
		}
		if (bAllNeighborsLand)
		{
			Frontier.RemoveSwap(FrontierIdx);
		}

		// Safety: prevent infinite loop
		if (Iteration > TargetLandCount * 20)
		{
			break;
		}
	}

	// Create water border tiles (1 ring around land for rendering)
	TSet<FHexCoord> WaterCoords;
	for (const int32 LandIdx : LandIndices)
	{
		TArray<FHexCoord> Nbs = Tiles[LandIdx].Coord.Neighbors();
		for (const FHexCoord& Nb : Nbs)
		{
			if (!CoordToIndex.Contains(Nb))
			{
				WaterCoords.Add(Nb);
			}
			else if (!Tiles[CoordToIndex[Nb]].bIsLand)
			{
				// Already exists as water, fine
			}
		}
	}

	for (const FHexCoord& WCoord : WaterCoords)
	{
		GetOrCreateTile(WCoord); // bIsLand defaults to false
	}
}

File: web-sandbox/src/sim/hexgrid.ts
import { HexGrid, HexTile } from './types.js';
import { rand01, chooseIndex } from './rng.js';

// Flat-top hex.
// World coords: wx = size * 1.5 * q,  wy = size * sqrt(3) * (r + q*0.5)
// Corner i (flat-top): angle = 60° * i  →  (cos, sin)
// Neighbor directions per edge index 0..5 (edge i = border shared with neighbor i):
//   0:(+1, 0)  1:(0,+1)  2:(-1,+1)  3:(-1,0)  4:(0,-1)  5:(+1,-1)

export const NEIGHBOR_DQ = [ 1,  0, -1, -1,  0,  1];
export const NEIGHBOR_DR = [ 0,  1,  1,  0, -1, -1];

export function coordKey(q: number, r: number): string { return `${q},${r}`; }

export function hexWorldPos(q: number, r: number, size: number): [number, number] {
  return [size * 1.5 * q, size * Math.sqrt(3) * (r + q * 0.5)];
}

export function hexCorners(cx: number, cy: number, size: number): [number, number][] {
  const corners: [number, number][] = [];
  for (let i = 0; i < 6; i++) {
    const a = (Math.PI / 3) * i;
    corners.push([cx + size * Math.cos(a), cy + size * Math.sin(a)]);
  }
  return corners;
}

export function hexDistance(q1: number, r1: number, q2: number, r2: number): number {
  const dq = q1 - q2, dr = r1 - r2;
  return Math.max(Math.abs(dq), Math.abs(dr), Math.abs(dq + dr));
}

function getOrCreate(grid: HexGrid, q: number, r: number): number {
  const key = coordKey(q, r);
  let idx = grid.coordMap.get(key);
  if (idx !== undefined) return idx;
  const [wx, wy] = hexWorldPos(q, r, grid.hexSize);
  idx = grid.tiles.length;
  const tile: HexTile = { index: idx, q, r, wx, wy, isLand: false, regionId: -1 };
  grid.tiles.push(tile);
  grid.coordMap.set(key, idx);
  return idx;
}

export function generateContinent(targetLandCount: number, seed: string, hexSize: number): HexGrid {
  const grid: HexGrid = { tiles: [], coordMap: new Map(), landIndices: [], hexSize };

  const centerIdx = getOrCreate(grid, 0, 0);
  grid.tiles[centerIdx].isLand = true;
  grid.landIndices.push(centerIdx);

  const frontier: number[] = [centerIdx];
  const maxRadius = Math.ceil(Math.sqrt(targetLandCount) * 1.2);
  let iter = 0;

  while (grid.landIndices.length < targetLandCount && frontier.length > 0) {
    const fi = chooseIndex(seed, `continent|pick|${iter}`, frontier.length);
    const tileIdx = frontier[fi];
    const { q, r } = grid.tiles[tileIdx];

    const nbEdge = chooseIndex(seed, `continent|nb|${iter}`, 6);
    const nq = q + NEIGHBOR_DQ[nbEdge];
    const nr = r + NEIGHBOR_DR[nbEdge];

    iter++;
    const dist = hexDistance(0, 0, nq, nr);
    if (dist > maxRadius + 2) continue;

    const acceptChance = Math.max(0.1, Math.min(1.0, 1 - dist / maxRadius));
    if (rand01(seed, `continent|accept|${iter}`) > acceptChance) continue;

    const nbIdx = getOrCreate(grid, nq, nr);
    if (!grid.tiles[nbIdx].isLand) {
      grid.tiles[nbIdx].isLand = true;
      grid.landIndices.push(nbIdx);
      frontier.push(nbIdx);
    }

    // Remove frontier tile if all 6 neighbors are land
    let allLand = true;
    for (let e = 0; e < 6; e++) {
      const nnIdx = grid.coordMap.get(coordKey(q + NEIGHBOR_DQ[e], r + NEIGHBOR_DR[e]));
      if (nnIdx === undefined || !grid.tiles[nnIdx].isLand) { allLand = false; break; }
    }
    if (allLand) frontier.splice(fi, 1);

    if (iter > targetLandCount * 20) break;
  }

  // Create 1-ring water border for neighbor lookups
  const waterToCreate = new Set<string>();
  for (const li of grid.landIndices) {
    const { q, r } = grid.tiles[li];
    for (let e = 0; e < 6; e++) {
      const k = coordKey(q + NEIGHBOR_DQ[e], r + NEIGHBOR_DR[e]);
      if (!grid.coordMap.has(k)) waterToCreate.add(k);
    }
  }
  for (const k of waterToCreate) {
    const [qs, rs] = k.split(',').map(Number);
    getOrCreate(grid, qs, rs); // isLand defaults to false
  }

  return grid;
}
